OpenSim/Simulation/SimbodyEngine/Joint.cpp

```diff
@@ -266,43 +266,30 @@ bool Joint::isCoordinateUsed(const Coordinate& aCoordinate) const
 
 void Joint::scale(const ScaleSet& scaleSet)
 {
-    SimTK::Vec3 parentFactors(1.0);
-    SimTK::Vec3 childFactors(1.0);
-
-    // Find the factors associated with the PhysicalFrames this Joint connects
-    const string& parentName = getParentFrame().findBaseFrame().getName();
-    const string& childName = getChildFrame().findBaseFrame().getName();
-    // Get scale factors
-    bool found_p = false;
-    bool found_b = false;
-    for (int i = 0; i < scaleSet.getSize(); i++) {
-        Scale& scale = scaleSet.get(i);
-        if (!found_p && (scale.getSegmentName() == parentName)) {
-            scale.getScaleFactors(parentFactors);
-            found_p = true;
-        }
-        if (!found_b && (scale.getSegmentName() == childName)) {
-            scale.getScaleFactors(childFactors);
-            found_b = true;
+    auto scaleOffsetFrameIfOwned = [&](const PhysicalFrame& frame) -> void {
+        // If the frame is owned by this Joint then scale it.
+        // Otherwise let the owner of the frame decide.
+        if (auto* offset = findComponent<PhysicalOffsetFrame>(frame.getName()))
```

Comment:
> I don't think it really matters where the frame lives inside the `Joint` as long as it is the *child* or *parent* frame. If it is some other frame it should not get scaled in this scheme.

Let *F* be a Frame owned by a Joint, where *F* is neither the Joint's child nor parent frame. Where does *F* get scaled "in this scheme"?

Reply:
> Let F be a Frame owned by a Joint, where F is neither the Joint's child nor parent frame. Where does F get scaled "in this scheme"?

*F* does not get scaled by the `Joint` in this scheme. It is not an attempt at a solution to scaling, which is a bigger issue. In the existing scheme (which this PR does not change), the Joint has jurisdiction over the scaling of *parent* and *child* frames that are PhysicalOffsetFrames. There is no change to that.

A better scheme may be for `Joint` to have no say in the scaling of frames whatsoever and that require the *base* frame to scale `PhysicalOffsetFrame`s attached to it, regardless of ownership. Let's work out a better scheme for post 4.0.

Reply:
> I don't think it really matters where the frame lives inside the `Joint` as long as it is the *child* or *parent* frame. If it is some other frame it should not get scaled in this scheme.

I think there's still an issue here that will result in some PhysicalOffsetFrames being scaled 0 times, some once, and some twice. I believe you're advocating that the frames that will be scaled 0 times should be ignored and addressed post-4.0. I think these situations should be detected and either throw an exception or at least issue a warning to the user (see suggestion added to Issue #1923).

I think the following scenario will result in a PhysicalOffsetFrame being scaled twice. Let *F* be a Frame and let *J1* and *J2* be Joints. Suppose *J1* resides in one of Model's subcomponent lists and *J2* resides in *J1*'s subcomponent list. Further suppose that *F* resides in one of *J2*'s subcomponent lists (either "frames" or "components") and that *F* is the parent or child frame for both *J1* and *J2*. I think the `findComponent()` call on L272 will cause both *J1* and *J2* to scale *F* because it is a descendant of both. Also fine if you want to push that to post-4.0, but I think it's a disservice to the user if we allow them to create models with arbitrary ownership trees (and, in fact, I believe that is one of the selling points of 4.0) without being able to scale or otherwise process such models. IMHO there at least needs to be a "scalability checker" that runs through all PhysicalOffsetFrames in the model and tells the user if any frames won't get scaled.